File: code/018/python/main.py
"""From Python to Go: Python: 017 - NETCONF."""

# Modules
import argparse
import datetime
import os
import sys
from dataclasses import dataclass
from typing import List
import difflib
import pygnmi.client
import yaml
import pygnmi
# from scrapli_netconf.driver import NetconfDriver
# import xmltodict


# Classes
@dataclass
class Credentials:
    """Class to store credentials."""
    username: str
    password: str


@dataclass
class Instruction:
    """Class to store instructions."""
    command: List[str]
    config: List[tuple]


@dataclass
class Result:
    """Class to store command execution data."""
    instruction: Instruction
    diff: str
    timestamp: datetime.datetime


class Device:
    """Class to interact with netowrk device."""
    def __init__(self, hostname: str, ip_address: str, port: int, platform: str, credentials: Credentials):
        self.hostname = hostname
        self.ip_address = ip_address
        self.port = port
        self.platform = platform
        self.credentials = credentials

        self.results: List[Result] = []

    def execute_change(self, instruction: Instruction) -> None:
        """Method to execute change."""

        # Connect to device
        with pygnmi.client.gNMIclient(
            target=(self.ip_address, self.port),
            username=self.credentials.username,
            password=self.credentials.password,
            skip_verify=True,
            timeout=5,
        ) as gconn:
            # Get state before change
            before = gconn.get(path=instruction.command, datatype="config")
            before_stringified = self.dict_to_xpath(before)

            # Apply change
            config_result = gconn.set(update=instruction.config, encoding="json_ietf")
            print(f"{config_result=}")

            # Get state after change
            after = gconn.get(path=instruction.command, datatype="config")
            after_stringified = self.dict_to_xpath(after)

            # Diff
            diff = "\n".join(
                difflib.context_diff(
                    before_stringified,
                    after_stringified,
                    lineterm="",
                )
            )

            self.results.append(
                Result(
                    instruction=instruction,
                    diff=diff,
                    timestamp=datetime.datetime.now(),
                )
            )

    def dict_to_xpath(self, data: dict) -> list:
        """Method to convert dict to xpath."""
        result = []

        if isinstance(data, str):
            return data

        for key, value in data.items():
            if isinstance(value, list):
                for ind, item in enumerate(value):
                    tr = self.dict_to_xpath(item)
                    result.extend([f"{key}/{ind}/{_}" for _ in tr])

            elif isinstance(value, dict):
                tr = self.dict_to_xpath(value)
                result.extend([f"{key}/{_}" for _ in tr])

            else:
                result.append(f"{key} = {value}")

        return result


# Functions
def read_args() -> argparse.Namespace:
    """Helper function to read CLI arguments."""
    parser = argparse.ArgumentParser(description="User input.")
    parser.add_argument("-i", "--inventory", type=str, help="Path to inventory file.")
    return parser.parse_args()


def load_inventory(filename: str, credentials: Credentials) -> List[Device]:
    """Function to load inventory data."""
    # Open file
    try:
        with open(filename, "r", encoding="utf-8") as f:
            data = yaml.safe_load(f)

    except FileNotFoundError as e:
        print(e)
        sys.exit(1)

    # Populate list of devices
    result = []
    for device in data:
        result.append(Device(credentials=credentials, **device))

    return result


def get_credentials() -> Credentials:
    """Function to get credentials."""
    username = os.getenv("AUTOMATION_USER")
    password = os.getenv("AUTOMATION_PASS")
    return Credentials(username, password)


# Main code
if __name__ == "__main__":
    # Read CLI arguments
    args = read_args()

    # Get credentials
    credentials = get_credentials()

    # Load inventory
    devices = load_inventory(args.inventory, credentials=credentials)

    # Config
    instruction = Instruction(
        command=["/openconfig-interfaces:interfaces"],
        config=[
            (
                "/openconfig-interfaces:interfaces",
                {
                    "interface": [
                        {
                            "name": "Loopback 23",
                            "config": {
                                "name": "Loopback 23",
                                "description": "Test-gnmi-python-2",
                            }
                        },
                    ],
                },
            ),
        ],
    )

    # Execute command
    for device in devices:
        device.execute_change(instruction)

    # Print results
    for device in devices:
        print(f"Device: {device.hostname}")
        for result in device.results:
            print(f"Config: {result.instruction.config}", f"Impact: {result.diff}", f"Timestamp: {result.timestamp}", sep="\n")

File: code/018/go/main.go
/* From Python to Go: Go: 018 - GNMI. */

package main

// Imports
import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/google/go-cmp/cmp"
	"github.com/openconfig/gnmic/pkg/api"
	"google.golang.org/protobuf/encoding/prototext"
	"gopkg.in/yaml.v3"
)

// Types and Receivers
type Arguments struct {
	/* Class to starte CLI arguments */
	Inventory string
}

type Crendetials struct {
	/* Struct to store credentials. */
	Username string
	Password string
}

type Instruction struct {
	Command string
	Config  struct {
		Path  string
		Value any
	}
}

type Result struct {
	/* Struct to store command execution result. */
	Instruction Instruction
	Diff        string
	Timestamp   time.Time
}

type Device struct {
	/* Struct to interact with netowrk device. */
	Hostname    string `yaml:"hostname"`
	IpAddress   string `yaml:"ip_address"`
	Port        uint   `yaml:"port"`
	Platform    string `yaml:"platform"`
	Crendetials Crendetials
	Result      []Result
}

func (d *Device) executeChange(i Instruction) {
	/* Method to execute command */

	// Create GNMI Target
	gnmiTarget, err := api.NewTarget(
		api.Name(d.Hostname),
		api.Address(fmt.Sprintf("%s:%d", d.IpAddress, d.Port)),
		api.Username(d.Crendetials.Username),
		api.Password(d.Crendetials.Password),
		api.SkipVerify(true),
	)
	if err != nil {
		log.Fatal("Cannot create GNMI Target: ", err)
	}

	// Create context
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Create GNMI client
	err = gnmiTarget.CreateGNMIClient(ctx)
	if err != nil {
		log.Fatal("Cannot create GNMI Client: ", err)
	}
	defer gnmiTarget.Close()

	// Get state before change
	getReq, err := api.NewGetRequest(
		api.Path(i.Command),
		api.DataType("config"),
		api.Encoding("json_ietf"),
	)
	if err != nil {
		log.Fatal("Cannot create Get request: ", err)
	}
	beforeGetResponse, err := gnmiTarget.Get(ctx, getReq)
	if err != nil {
		log.Fatal("Cannot make a Get request: ", err)
	}
	beforeStruct := OpenConfigInterfaces{}
	err = json.Unmarshal(beforeGetResponse.Notification[0].Update[0].Val.GetJsonIetfVal(), &beforeStruct)
	if err != nil {
		log.Fatal("Cannot unmarshall JSON: ", err)
	}

	// Make change
	setReq, err := api.NewSetRequest(
		api.Update(
			api.Path(i.Config.Path),
			api.Value(i.Config.Value, "json_ietf"),
		),
	)
	if err != nil {
		log.Fatal("Cannot create Set request: ", err)
	}
	setResp, err := gnmiTarget.Set(ctx, setReq)
	if err != nil {
		log.Fatal("Cannot make a Set request: ", err)
	}
	log.Println(prototext.Format(setResp))

	// Get state after change
	afterGetResponse, err := gnmiTarget.Get(ctx, getReq)
	if err != nil {
		log.Fatal("Cannot make a Get request: ", err)
	}
	afterStruct := OpenConfigInterfaces{}
	err = json.Unmarshal(afterGetResponse.Notification[0].Update[0].Val.GetJsonIetfVal(), &afterStruct)
	if err != nil {
		log.Fatal("Cannot unmarshall JSON: ", err)
	}
	// Diff
	diff := cmp.Diff(beforeStruct, afterStruct)

	// Update the result
	(*d).Result = append((*d).Result, Result{
		Instruction: i,
		Diff:        diff,
		Timestamp:   time.Now(),
	})
}

// Functions
func readArgs() Arguments {
	/* Helper function to read CLI arguments */
	result := Arguments{}

	flag.StringVar(&result.Inventory, "i", "", "Path to the inventory file")

	flag.Parse()

	return result
}

func loadInventory(p string) *[]Device {
	/* Function to load inventory data. */

	// Open file
	bs, err := os.ReadFile(p)
	if err != nil {
		fmt.Println("Get error ", err)
		os.Exit(1)
	}

	// Load inventory
	result := &[]Device{}

	err = yaml.Unmarshal(bs, result)
	if err != nil {
		fmt.Println("Get error ", err)
		os.Exit(1)
	}

	// Return result
	return result
}

func getCredentials() Crendetials {
	/* Function to get credentials. */
	return Crendetials{
		Username: os.Getenv("AUTOMATION_USER"),
		Password: os.Getenv("AUTOMATION_PASS"),
	}
}

// Main
func main() {
	/* Core logic */
	// Read CLI arguments
	cliArgs := readArgs()

	// Get credentials
	sshCreds := getCredentials()

	// Load inventory
	inventory := loadInventory(cliArgs.Inventory)

	// Config
	instruction := Instruction{
		Command: "/openconfig-interfaces:interfaces",
		Config: struct {
			Path  string
			Value any
		}{
			Path: "/openconfig-interfaces:interfaces",
			Value: map[string]any{
				"interface": []map[string]any{
					{
						"name": "Loopback 23",
						"config": map[string]any{
							"name":        "Loopback 23",
							"description": "Test-gnmi-golang-3",
						},
					},
				},
			},
		},
	}

	// Execute commands
	for i := 0; i < len(*inventory); i++ {
		(*inventory)[i].Crendetials = sshCreds
		(*inventory)[i].executeChange(instruction)
	}

	// Print results
	for i := 0; i < len(*inventory); i++ {
		for j := 0; j < len((*inventory)[i].Result); j++ {
			fmt.Printf(
				"Config: %v\nImpact: %v\nTimestamp: %v\n",
				(*inventory)[i].Result[j].Instruction.Config,
				(*inventory)[i].Result[j].Diff,
				(*inventory)[i].Result[j].Timestamp,
			)
		}
	}
}
